hbase-client/src/main/java/org/apache/hadoop/hbase/client/MetricsConnection.java

```diff
@@ -54,6 +58,38 @@
 @InterfaceAudience.Private
 public class MetricsConnection implements StatisticTrackable {
 
+  static final Map<String, MetricsConnection> METRICS_INSTANCES =
+    new HashMap<String, MetricsConnection>();
+
+  static MetricsConnection getMetricsConnection(final AsyncConnection conn,
+    Supplier<ThreadPoolExecutor> batchPool, Supplier<ThreadPoolExecutor> metaPool) {
+    String scope = getScope(conn);
+    MetricsConnection metrics;
+    synchronized (METRICS_INSTANCES) {
```

Comment:
Consider a synchronized or concurrent map type instead.

Reply:
I am not sure if a synchronized or concurrent map can protect the entire block, I want this entire block being single thread. Especially in the deletion method below, the decrementing count, getting count, and remove it from the map have to run in single thread mode. @apurtell @d-c-manning

Reply:
The scope here should be unique per async connection object, is that correct?

Reply:
I guess the scope that you meant here is the scope of this code block, it is per async connection object, and the single metrics object which might be shared among multiple async connection objects. Please refer the comment in the deleteMetricsConnection() code block, too.